File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/lib/UniffiCallbackGuard.kt
package io.nativeblocks.runtime.lib

import io.nativeblocks.runtime.engine.ErrorType
import io.nativeblocks.runtime.engine.NbException

internal inline fun <R> uniffiCallbackGuard(errorType: ErrorType, block: () -> R): R = try {
    block()
} catch (e: NbException) {
    throw e
} catch (e: Throwable) {
    throw NbException.Failure(
        reason = e.message ?: e.toString(),
        errorType = errorType,
        errorCode = null,
    )
}

File: platforms/ios/runtime/Sources/NativeblocksRuntime/lib/UniffiCallbackGuard.swift
import Foundation
import NativeblocksRuntimeFFI

internal func uniffiCallbackGuard<R>(_ errorType: ErrorType, _ block: () throws -> R) throws -> R {
    do {
        return try block()
    } catch let error as NbError {
        throw error
    } catch {
        throw NbError.Failure(
            reason: error.localizedDescription,
            errorType: errorType,
            errorCode: nil
        )
    }
}

internal func uniffiCallbackGuard<R>(_ errorType: ErrorType, _ block: () async throws -> R) async throws -> R {
    do {
        return try await block()
    } catch let error as NbError {
        throw error
    } catch {
        throw NbError.Failure(
            reason: error.localizedDescription,
            errorType: errorType,
            errorCode: nil
        )
    }
}
